src/components/files/fileHandler.ts

```diff
@@ -0,0 +1,119 @@
+import path from 'path'
+import fs from 'fs'
+import util from 'util'
+import { pipeline } from 'stream'
+import { nanoid } from 'nanoid'
+import { getAppRootDir } from '../utilityFunctions'
+import * as config from '../../config.json'
+import DBConnect from '../databaseConnect'
+import createThumbnail from './createThumbnails'
+const sharp = require('sharp')
+
+export const { filesFolderName } = config
+export const filesPath = path.join(getAppRootDir(), filesFolderName)
+
+interface HttpQueryParameters {
+  [key: string]: string
+}
+
+export function createFilesFolder() {
+  try {
+    fs.mkdirSync(path.join(getAppRootDir(), filesFolderName))
+  } catch {
+    // Folder already exists
+  }
+  // Move generic thumbnails to files root
+  fs.readdir(path.join(getAppRootDir(), 'images', 'generic_file_thumbnails'), (_, files) => {
+    files.forEach((file) =>
+      fs.copyFile(
+        path.join(getAppRootDir(), 'images', 'generic_file_thumbnails', file),
+        path.join(getAppRootDir(), filesFolderName, file),
+        () => {}
+      )
+    )
+  })
+}
+
+export async function getFilePath(uid: string, thumbnail = false) {
+  const result = await DBConnect.getFileDownloadInfo(uid, thumbnail)
+  if (!result) throw new Error()
+  return result
+}
+
+const pump = util.promisify(pipeline)
+
+export async function saveFiles(data: any, queryParams: HttpQueryParameters) {
+  const filesInfo = []
+  try {
+    for await (const file of data) {
```

Comment:
I am not entirely sure why await is necessary here, is it to await fo something inside for loop to finish (i think that would automatically await), is it to await for stream to be opened ? What happens if it's removed ?

Reply:
I'm don't fully understand it, but it doesn't work without it. 🤷‍♂️

There's some info on `for await...of` here : https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of

Without the await, the files don't even get written, so it must be something to do with what happens with the `await pump...... createFileStream`, I don't know, I just got the boilerplate from the official repo, and they use the `await`: https://github.com/fastify/fastify-multipart#handle-multiple-file-streams

Reply:
Ah, I think it's because `data` is an async Promise, so the await makes the loop wait for `data` to resolve before iterating over it.

Reply:
cool, makes sense, would you mind adding a comment ?

Reply:
Sure. 6643e49